text/0004-api-history-schema.md

```diff
@@ -0,0 +1,216 @@
+- Start Date: 2024-06-11
+- RFC PR: [electron/rfcs#6](https://github.com/electron/rfcs/pull/6)
+- Electron Issues: Related to
+[electron/electron#39877](https://github.com/electron/electron/issues/39877)
+- Reference Implementation:
+[electron/lint-roller (feat/api-history-schema-rfc)](https://github.com/electron/lint-roller/commit/d88115ad9ef0df7966976220adb7846f4338caf3)
+- Status: **Proposed**
+
+# API History Schema
+
+## Summary
+
+> [!NOTE]
+> This RFC is part of the **2024 Electron GSoC**
+> [[Project](https://summerofcode.withgoogle.com/programs/2024/projects/wenKR2i3)]
+> [[Proposal](https://gist.github.com/piotrpdev/9ed9bd7f0f3ab192a5ae58a35fbe03e7)].
+>
+> This RFC is **just for the schema**, not the API History feature as a whole.
+
+This proposal aims to introduce a
+[JSON Schema](https://json-schema.org/overview/what-is-jsonschema)
+for (YAML) API history blocks in `electron/electron` documentation.
+It will be used to enforce and validate the structure and content of these blocks,
+which will include historical information about changes to the various APIs in Electron.
+
+## Motivation
+
+The API history blocks need to be written in a well-designed format in order to
+convey information efficiently and ensure readability. These blocks will also be
+transformed into tables in `electron/website` to allow users to see what/when/how
+Electron APIs have changed. In order to do so, the blocks need to have a consistent
+and parsable structure.
+
+## Guide-level explanation
+
+Before this RFC, when a developer created a new feature they added documentation
+for it like this:
+
+`electron/docs/api/browser-window.md`
+
+```markdown
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with `titleBarStyle` set to `hidden`.
+```
+
+After this RFC, they will consult the schema, and also add an API history block:
+
+````markdown
+#### `win.setTrafficLightPosition(position)` _macOS_
+
+<!--
+```YAML history
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+```
+-->
+
+* `position` [Point](structures/point.md)
+
+Set a custom position for the traffic light buttons. Can only be used with
+`titleBarStyle` set to `hidden`.
+````
+
+Which, after several PRs, may reach this state:
+
+```yaml
+added:
+  - pr-url: https://github.com/electron/electron/pull/22533
+changes:
+  - pr-url: https://github.com/electron/electron/pull/26789
+    description: Made `trafficLightPosition` option work for `customButtonOnHover` window.
+deprecated:
+  - pr-url: https://github.com/electron/electron/pull/37094
+    breaking-changes-header: deprecated-browserwindowsettrafficlightpositionposition
+removed:
+  - pr-url: https://github.com/electron/electron/pull/39479
+    breaking-changes-header: removed-browserwindowsettrafficlightpositionposition
+```
```

Comment:
This is super nice! I like the consistency in structure. If one of the main advantages of the history is to help users upgrade their electron versions, would it be useful to explicitly add the affected Electron version(s) for each change without digging into the PR?

Reply:
That idea was part of my GSoC proposal, but @dsanders11 suggested deriving version numbers from the PR's themselves:
 
> _"...There’s one somewhat significant change we’d like to call out about the proposal, which came up during discussion when we were reviewing proposals. [...] [we] decided that the approach with the least drawbacks would be to only use PR URLs (the root PRs to main) instead of hardcoded version strings as in the proposal. This can serve as a single source of truth which can then be used to derive exact version numbers, and no further documentation changes on main are necessary if the change is backported to other branches."_ — @dsanders11 via Slack

I also mentioned this a little lower down in the RFC:

https://github.com/electron/rfcs/pull/6/files#diff-6f2f0631f8ae0584fb9798327b56c1ae7c89481e696e38dc6717c1cb8e14af5cR149-R152

```markdown
- The website will use the PR URLs to determine and display Electron version
numbers for the changes. This is more efficient than including the version
numbers in the Markdown docs themselves. It also removes the need to manually
update the docs when a backport is made.
```